File: Dev/unitTest_Engine_cpp_gtest/ObjectSystem/LifeCycle.cpp
#include <gtest/gtest.h>
#include <Altseed.h>
#include "../EngineTest.h"

using namespace std;
using namespace asd;

class ObjectSystem_LifeCycle : public EngineTest
{
	class MessageObject : public TextureObject2D
	{
	protected:
		void OnAdded() override
		{
			isAdded = true;
		}

		void OnRemoved() override
		{
			isAdded = false;
		}

		void OnDispose() override
		{
			haveBeenDisposed = true;
		}

		void OnUpdate() override
		{
			SetAngle(GetAngle() + 2);
		}

		void OnDrawAdditionally() override
		{
			DrawCircleAdditionally(GetPosition(), 50, 10, Color(0, 255, 0, 255), 12, 0, nullptr, AlphaBlendMode::Blend, 2);
		}

	public:
		bool isAdded, haveBeenDisposed;

		MessageObject()
			: isAdded(false)
			, haveBeenDisposed(false)
		{
			auto texture = Engine::GetGraphics()->CreateTexture2D(ToAString("Data/Textures/Cloud1.png").c_str());
			SetTexture(texture);
		}
	};

	class MessageLayer : public Layer2D
	{
	protected:
		void OnAdded() override
		{
			isAdded = true;
		}

		void OnRemoved() override
		{
			isAdded = false;
		}

		void OnUpdating() override
		{
			counter++;
		}

		void OnUpdated() override
		{
			if (counter % 30 == 0)
			{
				AddObject(make_shared<MessageObject>());
			}
		}

		void OnDrawAdditionally() override
		{
			DrawCircleAdditionally(Vector2DF(100, 100), 100, 10, Color(255, 0, 0, 255), 12, 0, nullptr, AlphaBlendMode::Blend, 2);
		}

		void OnVanish() override
		{
			ACE_ASSERT(false, "OnVanish is Obsolete.");
		}

		void OnDispose() override
		{
			haveBeenDisposed = true;
		}

	public:
		bool isAdded, haveBeenDisposed;
		int counter;

		MessageLayer()
			: isAdded(false)
			, haveBeenDisposed(false)
			, counter(0)
		{
		}
	};

	class MessageScene : public Scene
	{
	protected:
		void OnRegistered() override
		{
			ASSERT_TRUE(phase == 0 || phase == 6);
			phase = 1;
		}

		void OnStartUpdating() override
		{
			ASSERT_EQ(phase, 1);
			phase = 2;
		}

		void OnTransitionFinished() override
		{
			ASSERT_EQ(phase, 2);
			phase = 3;
		}

		void OnChanging() override
		{
			ASSERT_EQ(phase, 3);
			phase = 4;
		}

		void OnStopUpdating() override
		{
			ASSERT_EQ(phase, 4);
			phase = 5;
		}

		void OnUnregistered() override
		{
			ASSERT_EQ(phase, 5);
			phase = 6;
		}

		void OnDispose() override
		{
			haveBeenDisposed = true;
		}

	public:
		int phase;
		bool haveBeenDisposed;

		MessageScene()
			: phase(0)
			, haveBeenDisposed(false)
		{
		}
	};

public:
	ObjectSystem_LifeCycle(bool isOpenGLMode)
		: EngineTest(asd::ToAString("LifeCycle"), isOpenGLMode, 120)
	{
	}

private:
	shared_ptr<MessageScene> scene;
	shared_ptr<MessageLayer> layer;
	shared_ptr<MessageObject> object;

protected:
	void OnStart()
	{
		scene = make_shared<MessageScene>();
		layer = make_shared<MessageLayer>();
		object = make_shared<MessageObject>();

		ASSERT_EQ(scene->phase, 0);
		ASSERT_EQ(scene->haveBeenDisposed, false);
		ASSERT_EQ(layer->haveBeenDisposed, false);
		ASSERT_EQ(layer->isAdded, false);
		ASSERT_EQ(object->isAdded, false);
		ASSERT_EQ(object->haveBeenDisposed, false);

		layer->AddObject(object);
		ASSERT_EQ(object->isAdded, true);
		layer->RemoveObject(object);
		ASSERT_EQ(object->isAdded, false);
		layer->AddObject(object);

		scene->AddLayer(layer);
		ASSERT_EQ(layer->isAdded, true);
		scene->RemoveLayer(layer);
		ASSERT_EQ(layer->isAdded, false);
		scene->AddLayer(layer);

		Engine::ChangeSceneWithTransition(scene, make_shared<TransitionFade>(0.3f, 0.3f));
	}

	void OnUpdating()
	{
		if (m_currentTime == 65)
		{
			Engine::ChangeSceneWithTransition(nullptr, make_shared<TransitionFade>(0.3f, 0.3f));
		}
	}

	void OnUpdated()
	{
	}

	void OnFinish()
	{
		scene->Dispose();
		ASSERT_EQ(scene->phase, 6);
		ASSERT_EQ(scene->haveBeenDisposed, true);
		ASSERT_EQ(layer->haveBeenDisposed, true);
		ASSERT_EQ(object->haveBeenDisposed, true);
	}
};

ENGINE_TEST(ObjectSystem, LifeCycle)

File: Dev/unitTest_Engine_cs/ObjectSystem2D/LifeCycle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.ObjectSystem2D
{
	class LifeCycle : EngineTest
	{
		class MessageObject : TextureObject2D
		{
			public bool IsAdded { get; private set; }
			public bool HaveBeenDisposed { get; private set; }

			public MessageObject()
			{
				Texture = Engine.Graphics.CreateTexture2D(CloudTexturePath);
				IsAdded = false;
				HaveBeenDisposed = false;
			}

			protected override void OnAdded()
			{
				IsAdded = true;
			}

			protected override void OnRemoved()
			{
				IsAdded = false;
			}

			protected override void OnDispose()
			{
				HaveBeenDisposed = true;
			}

			protected override void OnUpdate()
			{
				Angle += 2;
			}

			protected override void OnDrawAdditionally()
			{
				DrawCircleAdditionally(Position, 50, 10, new Color(0, 255, 0), 12, 0, null, AlphaBlendMode.Blend, 2);
			}
		}

		class MessageLayer : Layer2D
		{
			public bool IsAdded { get; private set; }
			public bool HaveBeenDisposed { get; private set; }
			private int counter;

			public MessageLayer()
			{
				IsAdded = false;
				HaveBeenDisposed = false;
				counter = 0;
			}

			protected override void OnAdded()
			{
				IsAdded = true;
			}

			protected override void OnRemoved()
			{
				IsAdded = false;
			}

			protected override void OnDispose()
			{
				HaveBeenDisposed = true;
			}

			protected override void OnUpdating()
			{
				counter++;
			}

			protected override void OnUpdated()
			{
				if(counter % 30 == 0)
				{
					AddObject(new MessageObject() { Position = new Vector2DF(200, 0) });
				}
			}

			protected override void OnDrawAdditionally()
			{
				DrawCircleAdditionally(new Vector2DF(100, 100), 100, 10, new Color(255, 0, 0), 12, 0, null, AlphaBlendMode.Blend, 2);
			}
		}

		class MessageScene : Scene
		{
			public int Phase { get; private set; }
			public bool HaveBeenDisposed { get; set; }

			public MessageScene()
			{
				Phase = 0;
				HaveBeenDisposed = false;
			}

			protected override void OnRegistered()
			{
				Assert.IsTrue(Phase == 0 || Phase == 6);
				Phase = 1;
			}

			protected override void OnStartUpdating()
			{
				Assert.AreEqual(1, Phase);
				Phase = 2;
			}

			protected override void OnTransitionFinished()
			{
				Assert.AreEqual(2, Phase);
				Phase = 3;
			}

			protected override void OnTransitionBegin()
			{
				Assert.AreEqual(3, Phase);
				Phase = 4;
			}

			protected override void OnStopUpdating()
			{
				Assert.AreEqual(4, Phase);
				Phase = 5;
			}

			protected override void OnUnregistered()
			{
				Assert.AreEqual(5, Phase);
				Phase = 6;
			}

			protected override void OnDispose()
			{
				HaveBeenDisposed = true;
			}
		}

		private MessageObject obj;
		private MessageLayer layer;
		private MessageScene scene;

		public LifeCycle() : base(120)
		{
		}

		protected override void OnStart()
		{
			scene = new MessageScene();
			layer = new MessageLayer();
			obj = new MessageObject();

			Assert.AreEqual(0, scene.Phase);
			Assert.AreEqual(false, scene.HaveBeenDisposed);
			Assert.AreEqual(false, layer.IsAdded);
			Assert.AreEqual(false, layer.HaveBeenDisposed);
			Assert.AreEqual(false, obj.IsAdded);
			Assert.AreEqual(false, obj.HaveBeenDisposed);

			layer.AddObject(obj);
			Assert.AreEqual(true, obj.IsAdded);
			layer.RemoveObject(obj);
			Assert.AreEqual(false, obj.IsAdded);
			layer.AddObject(obj);

			scene.AddLayer(layer);
			Assert.AreEqual(true, layer.IsAdded);
			scene.RemoveLayer(layer);
			Assert.AreEqual(false, layer.IsAdded);
			scene.AddLayer(layer);

			Engine.ChangeSceneWithTransition(scene, new TransitionFade(0.3f, 0.3f));
		}

		protected override void OnUpdating()
		{
			if(Time == 65)
			{
				Engine.ChangeSceneWithTransition(null, new TransitionFade(0.3f, 0.3f));
			}
		}

		protected override void OnFinish()
		{
			scene.Dispose();
			Assert.AreEqual(6, scene.Phase);
			Assert.AreEqual(true, scene.HaveBeenDisposed);
			Assert.AreEqual(true, layer.HaveBeenDisposed);
			Assert.AreEqual(true, obj.HaveBeenDisposed);
		}
	}
}
